src/pydvl/value/results.py

```diff
@@ -0,0 +1,232 @@
+indexing abilities, and conversion to `pandas DataFrames
+<https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html>`_.
+"""
+import collections.abc
+from dataclasses import dataclass
+from enum import Enum
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Union,
+    cast,
+    overload,
+)
+
+import numpy as np
+
+from pydvl.utils import Dataset, SortOrder
+
+try:
+    import pandas  # Try to import here for the benefit of mypy
+except ImportError:
+    pass
+
+if TYPE_CHECKING:
+    from numpy.typing import NDArray
+
+__all__ = ["ValuationResult", "ValuationStatus"]
+
+
+class ValuationStatus(Enum):
+    Pending = "pending"
+    Converged = "converged"
+    MaxIterations = "maximum number of iterations reached"
+    Failed = "failed"
+
+
+@dataclass
+class ValueItem:
+    """The result of a value computation for one datum."""
+
+    index: np.int_
+    name: str
+    value: np.float_
+    stderr: Optional[np.float_]
+
+
+class ValuationResult(collections.abc.Sequence):
+    """Objects of this class hold the results of valuation algorithms.
+
+    Results can be sorted with :meth:`sort`. Note that sorting values affects
+    how iterators and the object itself as ``Sequence`` behave: ``values[0]``
+    returns a :class:`ValueItem` with the highest or lowest ranking point if
+    this object is sorted by descending or ascending value, respectively. If
+    unsorted, ``values[0]`` returns a ``ValueItem`` for index 0.
+
+    Similarly, ``iter(valuation_result)`` returns a generator of
+    :class:`ValueItem` in the order in which the object is sorted.
+
+    :param algorithm: The method used.
+    :param status: The end status of the algorithm.
+    :param values: An array of values, data indices correspond to positions in
+        the array.
+    :param stderr: An optional array of standard errors in the computation of
+        each value.
+    :param data_names: Names for the data points. Defaults to index numbers
+        if not set.
+    :param sort: Whether to sort the values. See above how this affects usage
+        as an iterable or sequence.
+
+    :raise ValueError: If data names and values have mismatching lengths.
+    """
+
+    _indices: "NDArray[np.int_]"
+    _values: "NDArray[np.float_]"
+    _data: Dataset
+    _names: "Union[NDArray[np.int_], NDArray[np.str_]]"
+    _stderr: "NDArray[np.float_]"
+    _algorithm: str  # TODO: BaseValuator
+    _status: ValuationStatus  # TODO: Maybe? BaseValuator.Status
+    _sort_order: Optional[SortOrder] = None
+
+    def __init__(
+        self,
+        algorithm: Callable,  # BaseValuator,
+        status: ValuationStatus,  # Valuation.Status,
+        values: "NDArray[np.float_]",
+        stderr: Optional["NDArray[np.float_]"] = None,
+        data_names: Optional[Sequence[str]] = None,
+        sort: Optional[SortOrder] = None,
```

Comment:
Wouldn't it be simpler to just use a boolean flag since there are only 2 possible options?

Reply:
Yes, but a boolean would feel rather weird for "ascending" vs "descending". It is a total PITA having to import the class and to type so much, I agree. I wanted to use the string enum values, but there is no way to automatically construct a literal from them that mypy accepts, and in the end that kind of defeats the purpose since user code won't necessarily use a type checker.

Reply:
Also, I think originally I thought that one may want to add different sorting methods, e.g. taking into account the stderr. But that's admittedly rather unlikely to be very useful.

Reply:
I guess I'll just use a literal and do some arg checking inside sort()

Reply:
Before this change, the results were sorted in an ascending order by default. What I'm trying to understand is what would be the use case for sorting in a different order or for not sorting at all?

Reply:
Being able to pick the order is just a convenience. Sort descending (resp. ascending) to iterate first over high (resp. low) value points. No sorting means you iterate over points using their indices. I have made this optional just to be more explicit about what's going on.

Reply:
I've experimented with Literals, but I don't like how it looks. It's a purely aesthetic thing and hence subject to debate, but since it's a rather minor issue and I need to wrap this up before next week's meeting madness, I'm going to resolve the conversation. But feel free to reference this thread in a new issue if you find the Enum too awkward.

Reply:
I still have to say YAGNI to this change. `ValuationResult` would be a simpler as just a container and not a container with extra functionality that can also be obtained by converting it to a pandas dataframe.